lib/travis/logs/drain_queue.rb

```diff
@@ -37,52 +46,127 @@ def subscribe
         @jobs_channel ||= amqp_conn.create_channel
       end
 
+      private def batch_size
+        @batch_size ||= Integer(logs_config[:drain_batch_size] || 0)
+      end
+
       private def amqp_conn
         @amqp_conn ||= Bunny.new(amqp_config).tap(&:start)
       end
 
       private def amqp_config
-        Travis.config.amqp.to_h
+        @amqp_config ||= Travis.config.amqp.to_h
+      end
+
+      private def logs_config
+        @logs_config ||= Travis.config.logs.to_h
+      end
+
+      private def batch_buffer
+        @batch_buffer ||= Concurrent::Map.new
+      end
+
+      private def flush_mutex
+        @flush_mutex ||= Mutex.new
+      end
+
+      private def build_periodic_flush_task
+        Concurrent::TimerTask.execute(
+          run_now: true,
+          execution_interval: logs_config[:drain_execution_interval],
+          timeout_interval: logs_config[:drain_timeout_interval]
+        ) do
+          Travis.logger.debug(
+            'triggering periodic flush',
+            interval: "#{logs_config[:drain_execution_interval]}s",
+            timeout: "#{logs_config[:drain_timeout_interval]}s"
+          )
+          flush_mutex.synchronize { flush_batch_buffer }
+        end
+      end
+
+      private def flush_batch_buffer
+        Travis.logger.info('flushing batch buffer', size: batch_buffer.size)
+        sample = {}
+        payload = []
+
+        batch_buffer.each_pair do |delivery_tag, entry|
+          sample[delivery_tag] = entry
+        end
+
+        sample.each_pair do |delivery_tag, entry|
+          payload.push(entry)
+
+          begin
+            batch_buffer.delete_pair(delivery_tag, entry)
+          rescue StandardError => e
+            Travis.logger.error(
+              'failed to delete pair from buffer',
+              error: e.inspect
+            )
+            payload.pop
+            next
+          end
+
+          begin
+            jobs_channel.ack(delivery_tag, true)
```

Comment:
Is `true` correct here? As far as I understand from [the documentation](http://www.rubydoc.info/github/ruby-amqp/bunny/Bunny/Channel#ack-instance_method), that will cause all unacked messages up to this message to be acked, which I don't think is what we want?